docs/guides/spmd_api.md

```diff
@@ -0,0 +1,429 @@
+---
+
+<!-- #region id="2-JYevEklHfB" -->
+# Scale up Flax modules with `pjit`
+
+In this guide, you will learn how to use JAX's `pjit` and `flax.linen.spmd` to scale up your Flax modules on multiple devices and hosts.
+
+Modern JAX provides [`jax.experimental.pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html) as a way to automatically compile and scale up JAX computations using the [SPMD](https://jax.readthedocs.io/en/latest/glossary.html?highlight=spmd#term-SPMD) model. You can think of `pjit` as [`jit`](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html) where one explicitly specificies how the input and output data should be partitioned across devices. See the [JAX guide](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html) for more information on pjit.
+
+Flax provides an interface for you to specify partitions for your module parameters when defining your module, and later use it for `pjit` compilation. You can also use logical axis annotations to decouple your model code and partition plan, in order to customize and experiment with different partition layouts more easily.
+<!-- #endregion -->
+
+<!-- #region id="eTGIfmDclxJ2" -->
+# Setup
+
+Install Flax from head and make imports as necessary. 
+
+Note that this Colab uses `--xla_force_host_platform_device_count=8` to emulate multiple devices on a CPU environment. You may also choose to run on a multi-device TPU environment following [this guide](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html#setup), in which case you should ignore the `os.environ` cell.
+<!-- #endregion -->
+
+```python id="e3d005e672d6"
+# TODO: replace this with `pip3 install flax` when 0.6.4 release rolls out.
+!pip3 install -qq "git+https://github.com/google/flax.git@main#egg=flax"
+```
+
+```python id="FJzB9vfp-OoK"
+import os
+os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'
+```
+
+```python id="9JJbxi0zHmn3"
+import functools
+import numpy as np
+import jax
+
+from jax import lax, random, numpy as jnp
+
+import flax
+from flax import struct, traverse_util, linen as nn
+from flax.linen import spmd
+from flax.core import freeze, unfreeze
+from flax.training import train_state, checkpoints
+
+import optax
+```
+
+<!-- #region id="XfuDTqt1mO6g" -->
+Import all the `pjit` related libraries. Note that they are still in the experimental package of JAX, so some API may change. In the near future, the plan is to move the `pjit` API out of the experimental status and merge it with the current `jax.jit` API.
+
+Start a device mesh using the 8 devices available. In this example, we set them as a `2x4` device mesh, same as the layout of TPU v3-8, and annotate each axis with a name. A typical way to annotate axis names is `('data', 'model')`: `'data'` being the mesh dimension used for data-parallel sharding of the batch dimension of inputs and activations, and the `'model'` being the mesh dimension used for sharding parameters of the model across devices.
```

Comment:
Thinking about this some more, if you have a `2x4` device and you annotate axis names with `('data', 'model')`, then this mean that if you partition over the `data` dimension you are effectively doing a two-way partition over the first dimension of the devices, and if you partition over the `model` dimension you partition it four way over the second dimension, right? It might be good to explain this explicitly, so the relation between the devices and the axis names becomes clearer.